Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,629 @@
+
+
+        Example
+        -------
+        >>> sampler = Sampler.random_with_replacement(5)
+        >>> print(sampler.get_samples(10))
+        [3 4 0 0 2 3 3 2 2 1]
+
+        >>> sampler = Sampler.random_with_replacement(4, [0.7,0.1,0.1,0.1])
+        >>> print(sampler.get_samples(10))
+        [0 1 3 0 0 3 0 0 0 0]
+        """
+
+        if prob == None:
+            prob = [1/num_indices] * num_indices
+
+        sampler = SamplerRandom(
+            num_indices=num_indices, sampling_type='random_with_replacement', replace=True, prob=prob, seed=seed)
+        return sampler
+
+    @staticmethod
+    def random_without_replacement(num_indices, seed=None):
+        """
+        Instantiates a sampler which outputs from a list of indices {0, 1, …, S-1}, with S=num_indices, uniformly randomly without replacement.
+
+        Parameters
+        ----------
+        num_indices: int
+            One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+        seed:int, default=None
+            Random seed for the  random number generator.  If set to None, the seed will be set using the current time. 
+
+        Returns
+        -------
+        A Sampler  that can be called with Sampler.next()  or next(Sampler) that samples randomly without replacement 
+        Example
+        -------
+        >>> sampler=Sampler.randomWithoutReplacement(7, seed=1)
+        >>> print(sampler.get_samples(16))
+        [6 2 1 0 4 3 5 1 0 4 2 5 6 3 3 2]
+
+        """
+
+        sampler = SamplerRandom(
+            num_indices=num_indices, sampling_type='random_without_replacement', replace=False,  seed=seed)
+        return sampler
+
+    @staticmethod
+    def from_function(num_indices, function, prob_weights=None):
+        """
+        Instantiates a sampler that wraps a function that takes an iteration number and selects from a list of indices {0, 1, …, S-1}.
+
+        Parameters
+        ----------
+        num_indices: int
+            One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+        function: A function that takes an in integer iteration number and returns an integer from {0, 1, …, S-1} with S=num_indices. 
+
+        prob_weights: list of floats of length num_indices that sum to 1. Default is [1/num_indices]*num_indices 
+            Consider that the sampler is called a large number of times this argument holds the expected number of times each index would be called,  normalised to 1. 
+
+
+        Note
+        -----
+        If your function involves a random number generator, then it may be easier to subclass the SamplerRandom class instead. 
+
+        Returns
+        -------
+        A Sampler that wraps a function and can be called with Sampler.next()  or next(Sampler) 
+
+
+        Example
```

Comment:
This is too long an example. I would write it using 'my_sampling_function' (or similar). And leave probabilities as None.